File: classes/DictionaryParser.cpp
#include "DictionaryParser.h"
#include "rapidjson/document.h"
#include "rapidjson/filereadstream.h"
#include <cstdio>
#include <iostream>

using namespace rapidjson;

void DictionaryParser::parseAndAddToTree(const std::string& filename, AVLTree& tree) {
    FILE* file = fopen(filename.c_str(), "rb");
    if (!file) {
        std::cerr << "Error: Unable to open JSON file: " << filename << std::endl;
        return;
    }

    char buffer[65536];
    FileReadStream inputStream(file, buffer, sizeof(buffer));

    Document document;
    document.ParseStream(inputStream);

    fclose(file);

    if (document.HasParseError()) {
        std::cerr << "Error: JSON parse error: " << document.GetParseError() << std::endl;
        return;
    }

    if (!document.IsObject()) {
        std::cerr << "Error: JSON root is not an object." << std::endl;
        return;
    }

    for (Value::ConstMemberIterator itr = document.MemberBegin(); itr != document.MemberEnd(); ++itr) {
        const std::string value = itr->name.GetString();
        const std::string key = itr->value.GetString();
        tree.insert(key, value);
    }
}

File: classes/DictionaryParser.js
const fs = require('fs');
const { AVLTree } = require('./AVLTree');

class DictionaryParser {
    static parseAndAddToTree(filename, tree) {
        try {
            const data = fs.readFileSync(filename, 'utf8');
            const jsonData = JSON.parse(data);

            Object.keys(jsonData).forEach(key => {
                const value = jsonData[key];
                tree.insert(value, key);
            });
        } catch (err) {
            console.error(`${err}`);
        }
    }

    static trimResults(wordSet) {
        if (wordSet.size <= 1) return wordSet;
        let outputSet = new Set();
        let min = 10;

        wordSet.forEach(word => {
            if (word.split('/').length - 1 < min) {
                min = word.split('/').length - 1;
                outputSet = new Set();
                outputSet.add(word);
            } else if (word.split('/').length - 1 === min) {
                outputSet.add(word);
            }
        });
        if (outputSet.size > 1) {
            let newSet = new Set();
            outputSet.forEach(word => {
                if (word.split('-').length === 1) newSet.add(word);
            });
            outputSet.forEach(word => {
                if (word.split('*').length !== 1) newSet.delete(word);
            });
            if (newSet.size >= 1) {
                return newSet;
            }
        }
        return outputSet;
    }
}

module.exports = DictionaryParser;
